app/ui/browser-modern/util/session-migrate.js

```diff
@@ -0,0 +1,54 @@
+/*
+Copyright 2016 Mozilla
+
+Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+this file except in compliance with the License. You may obtain a copy of the
+License at http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software distributed
+under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+CONDITIONS OF ANY KIND, either express or implied. See the License for the
+specific language governing permissions and limitations under the License.
+*/
+
+import { logger } from '../../../shared/logging';
+
+/**
+ * This module provides utility functions for performing forward migration
+ * on different versions of the app state.
+ */
+
+// Whenever app state shape changes, migration functions need to be added here.
+// Add the obsolete or missing record names as keys, and a migration function
+// taking the data to migrate as a basic Immutable.Map, and having to return it
+// in the new format.
+const MIGRATIONS = {
+  // e.g. migrate MyRecord from v1 to v2:
```

Comment:
Perhaps worth including how to sequence migrations.  That is, if you get `MyRecord_v1` but you expect v3, you have to update `MyRecord_v1` and make `MyRecord_v2` call `MyRecord_v1`, etc.